crates/matrix-sdk/src/widget/filter.rs

```diff
@@ -103,271 +103,325 @@ pub enum StateEventFilter {
+        Self::State(FilterInputState { event_type: event_type.into(), state_key })
     }
 }
 
+/// Filter input data that is used for a [`FilterInput::State`] filter.
 #[derive(Debug, Deserialize)]
-pub(super) struct MatrixEventFilterInput {
+pub struct FilterInputState {
     #[serde(rename = "type")]
-    pub(super) event_type: TimelineEventType,
-    pub(super) state_key: Option<String>,
-    pub(super) content: MatrixEventContent,
+    pub(super) event_type: StateEventType,
+    pub(super) state_key: String,
 }
 
+// Filter input message like:
 #[derive(Debug, Default, Deserialize)]
-pub(super) struct MatrixEventContent {
+pub(super) struct MessageLikeFilterEventContent {
     pub(super) msgtype: Option<String>,
 }
 
+#[derive(Debug)]
+pub struct FilterInputMessageLike {
+    pub(super) event_type: MessageLikeEventType,
+    pub(super) content: MessageLikeFilterEventContent,
+}
+
+/// Create a filter input based on [`AnyTimelineEvent`].
+/// This will create a [`FilterInput::State`] or [`FilterInput::MessageLike`]
+/// depending on the event type.
+impl TryFrom<Raw<AnyTimelineEvent>> for FilterInput {
+    type Error = serde_json::Error;
+    fn try_from(raw_event: Raw<AnyTimelineEvent>) -> Result<Self, Self::Error> {
+        // make sure `raw_event` actually was a timeline event
+        let timeline_event = raw_event.deserialize()?;
+        match timeline_event {
+            AnyTimelineEvent::MessageLike(message_like) => {
+                let msgtype = if let Some(AnyMessageLikeEventContent::RoomMessage(r)) =
+                    message_like.original_content()
+                {
+                    Some(r.msgtype().to_owned())
+                } else {
+                    None
+                };
+                let input_content = FilterInputMessageLike {
+                    event_type: message_like.event_type(),
+                    content: MessageLikeFilterEventContent { msgtype },
+                };
+                Ok(FilterInput::MessageLike(input_content))
+            }
+            AnyTimelineEvent::State(s) => Ok(FilterInput::State(FilterInputState {
+                event_type: s.event_type(),
+                state_key: s.state_key().to_owned(),
+            })),
+        }
+    }
+}
+
+impl From<&SendEventRequest> for FilterInput {
+    fn from(request: &SendEventRequest) -> Self {
+        match &request.state_key {
+            None => match request.event_type {
+                TimelineEventType::RoomMessage => {
+                    if let Some(msgtype) =
+                        serde_json::from_str::<MessageLikeFilterEventContent>(request.content.get())
+                            .unwrap_or_else(|e| {
+                                debug!("Failed to deserialize event content for filter: {e}");
+                                // Fallback to empty content is safe because there is no filter
+                                // that matches with it when it otherwise wouldn't.
```

Comment:
That matches with what?

Is it really safe to default to an empty content here?